File: C/priority_queue_utils/test_priority_queue.c
/**
 * @file test_priority_queue.c
 * @brief Unit tests for Priority Queue (Min-Heap) Implementation
 * 
 * @author AllToolkit
 * @date 2026-04-23
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <assert.h>
#include "priority_queue.h"

/* Test counters */
static int tests_run = 0;
static int tests_passed = 0;

#define TEST(name) do { \
    printf("  Testing: %s... ", #name); \
    tests_run++; \
    if (test_##name()) { \
        printf("PASSED\n"); \
        tests_passed++; \
    } else { \
        printf("FAILED\n"); \
    } \
} while(0)

/* Comparison function for integers */
static int cmp_int(const void *a, const void *b) {
    return *(const int *)a - *(const int *)b;
}

/* ========== Test Functions ========== */

int test_create_free(void) {
    PriorityQueue *pq = pq_create(sizeof(int), 0);
    if (pq == NULL) return 0;
    
    int result = (pq_size(pq) == 0) && 
                 (pq_is_empty(pq)) && 
                 (pq_capacity(pq) > 0);
    
    pq_free(&pq);
    return result && (pq == NULL);
}

int test_create_with_capacity(void) {
    PriorityQueue *pq = pq_create(sizeof(int), 100);
    if (pq == NULL) return 0;
    
    int result = (pq_capacity(pq) == 100) && (pq_size(pq) == 0);
    
    pq_free(&pq);
    return result;
}

int test_insert_extract_int(void) {
    PriorityQueue *pq = pq_create_int();
    if (pq == NULL) return 0;
    
    /* Insert in random order */
    pq_insert_int(pq, 30, 3.0);
    pq_insert_int(pq, 10, 1.0);
    pq_insert_int(pq, 20, 2.0);
    pq_insert_int(pq, 5, 0.5);
    
    if (pq_size(pq) != 4) {
        pq_free(&pq);
        return 0;
    }
    
    /* Extract in priority order */
    int val;
    double pri;
    
    pq_extract_int(pq, &val, &pri);
    if (val != 5 || pri != 0.5) {
        pq_free(&pq);
        return 0;
    }
    
    pq_extract_int(pq, &val, &pri);
    if (val != 10 || pri != 1.0) {
        pq_free(&pq);
        return 0;
    }
    
    pq_extract_int(pq, &val, &pri);
    if (val != 20 || pri != 2.0) {
        pq_free(&pq);
        return 0;
    }
    
    pq_extract_int(pq, &val, &pri);
    if (val != 30 || pri != 3.0) {
        pq_free(&pq);
        return 0;
    }
    
    int result = pq_is_empty(pq);
    pq_free(&pq);
    return result;
}

int test_peek(void) {
    PriorityQueue *pq = pq_create_int();
    if (pq == NULL) return 0;
    
    pq_insert_int(pq, 42, 2.0);
    pq_insert_int(pq, 99, 1.0);  /* Lower priority, should be at front */
    
    int val;
    double pri;
    if (!pq_peek_int(pq, &val, &pri) || val != 99 || pri != 1.0) {
        pq_free(&pq);
        return 0;
    }
    
    /* Peek should not remove element */
    if (pq_size(pq) != 2) {
        pq_free(&pq);
        return 0;
    }
    
    pq_free(&pq);
    return 1;
}

int test_same_priority(void) {
    PriorityQueue *pq = pq_create_int();
    if (pq == NULL) return 0;
    
    /* Insert elements with same priority */
    pq_insert_int(pq, 10, 1.0);
    pq_insert_int(pq, 20, 1.0);
    pq_insert_int(pq, 30, 1.0);
    
    /* All should be extractable */
    int val;
    int count = 0;
    while (!pq_is_empty(pq)) {
        pq_extract_int(pq, &val, NULL);
        count++;
    }
    
    pq_free(&pq);
    return count == 3;
}

int test_is_empty(void) {
    PriorityQueue *pq = pq_create_int();
    if (pq == NULL) return 0;
    
    if (!pq_is_empty(pq)) {
        pq_free(&pq);
        return 0;
    }
    
    pq_insert_int(pq, 1, 1.0);
    if (pq_is_empty(pq)) {
        pq_free(&pq);
        return 0;
    }
    
    int val;
    pq_extract_int(pq, &val, NULL);
    if (!pq_is_empty(pq)) {
        pq_free(&pq);
        return 0;
    }
    
    pq_free(&pq);
    return 1;
}

int test_clear(void) {
    PriorityQueue *pq = pq_create_int();
    if (pq == NULL) return 0;
    
    for (int i = 0; i < 100; i++) {
        pq_insert_int(pq, i, (double)i);
    }
    
    if (pq_size(pq) != 100) {
        pq_free(&pq);
        return 0;
    }
    
    pq_clear(pq);
    
    int result = (pq_size(pq) == 0) && pq_is_empty(pq);
    pq_free(&pq);
    return result;
}

int test_reserve(void) {
    PriorityQueue *pq = pq_create_int();
    if (pq == NULL) return 0;
    
    if (!pq_reserve(pq, 1000)) {
        pq_free(&pq);
        return 0;
    }
    
    if (pq_capacity(pq) != 1000) {
        pq_free(&pq);
        return 0;
    }
    
    /* Reserve smaller should not shrink */
    pq_reserve(pq, 500);
    if (pq_capacity(pq) != 1000) {
        pq_free(&pq);
        return 0;
    }
    
    pq_free(&pq);
    return 1;
}

int test_shrink_to_fit(void) {
    PriorityQueue *pq = pq_create_int();
    if (pq == NULL) return 0;
    
    pq_reserve(pq, 100);
    for (int i = 0; i < 10; i++) {
        pq_insert_int(pq, i, (double)i);
    }
    
    if (!pq_shrink_to_fit(pq)) {
        pq_free(&pq);
        return 0;
    }
    
    int result = (pq_capacity(pq) == 10) && (pq_size(pq) == 10);
    pq_free(&pq);
    return result;
}

int test_contains_priority(void) {
    PriorityQueue *pq = pq_create_int();
    if (pq == NULL) return 0;
    
    pq_insert_int(pq, 10, 1.0);
    pq_insert_int(pq, 20, 2.0);
    pq_insert_int(pq, 30, 3.0);
    
    if (!pq_contains_priority(pq, 1.0)) {
        pq_free(&pq);
        return 0;
    }
    
    if (!pq_contains_priority(pq, 2.0)) {
        pq_free(&pq);
        return 0;
    }
    
    if (pq_contains_priority(pq, 99.0)) {
        pq_free(&pq);
        return 0;
    }
    
    pq_free(&pq);
    return 1;
}

int test_update_priority(void) {
    PriorityQueue *pq = pq_create_int();
    if (pq == NULL) return 0;
    
    pq_insert_int(pq, 10, 3.0);
    pq_insert_int(pq, 20, 2.0);
    pq_insert_int(pq, 30, 1.0);
    
    /* Update priority of element 20 from 2.0 to 0.5 (should become min) */
    int target = 20;
    if (!pq_update_priority(pq, &target, 0.5, cmp_int)) {
        pq_free(&pq);
        return 0;
    }
    
    /* Verify 20 is now at front */
    int val;
    double pri;
    pq_peek_int(pq, &val, &pri);
    if (val != 20 || pri != 0.5) {
        pq_free(&pq);
        return 0;
    }
    
    pq_free(&pq);
    return 1;
}

int test_merge(void) {
    PriorityQueue *pq1 = pq_create_int();
    PriorityQueue *pq2 = pq_create_int();
    if (pq1 == NULL || pq2 == NULL) {
        if (pq1) pq_free(&pq1);
        if (pq2) pq_free(&pq2);
        return 0;
    }
    
    pq_insert_int(pq1, 10, 1.0);
    pq_insert_int(pq1, 30, 3.0);
    
    pq_insert_int(pq2, 20, 2.0);
    pq_insert_int(pq2, 40, 4.0);
    
    if (!pq_merge(pq1, pq2)) {
        pq_free(&pq1);
        pq_free(&pq2);
        return 0;
    }
    
    if (pq_size(pq1) != 4) {
        pq_free(&pq1);
        pq_free(&pq2);
        return 0;
    }
    
    /* Extract should be in sorted order */
    int val;
    double pri;
    pq_extract_int(pq1, &val, &pri);
    if (val != 10 || pri != 1.0) {
        pq_free(&pq1);
        pq_free(&pq2);
        return 0;
    }
    
    pq_extract_int(pq1, &val, &pri);
    if (val != 20 || pri != 2.0) {
        pq_free(&pq1);
        pq_free(&pq2);
        return 0;
    }
    
    pq_free(&pq1);
    pq_free(&pq2);
    return 1;
}

int test_get_sorted(void) {
    PriorityQueue *pq = pq_create_int();
    if (pq == NULL) return 0;
    
    pq_insert_int(pq, 30, 3.0);
    pq_insert_int(pq, 10, 1.0);
    pq_insert_int(pq, 20, 2.0);
    pq_insert_int(pq, 5, 0.5);
    
    int sorted[4];
    double priorities[4];
    
    if (!pq_get_sorted(pq, sorted, priorities)) {
        pq_free(&pq);
        return 0;
    }
    
    /* Verify sorted order */
    if (sorted[0] != 5 || priorities[0] != 0.5) {
        pq_free(&pq);
        return 0;
    }
    if (sorted[1] != 10 || priorities[1] != 1.0) {
        pq_free(&pq);
        return 0;
    }
    if (sorted[2] != 20 || priorities[2] != 2.0) {
        pq_free(&pq);
        return 0;
    }
    if (sorted[3] != 30 || priorities[3] != 3.0) {
        pq_free(&pq);
        return 0;
    }
    
    /* Original queue should be unchanged */
    if (pq_size(pq) != 4) {
        pq_free(&pq);
        return 0;
    }
    
    pq_free(&pq);
    return 1;
}

int test_double_type(void) {
    PriorityQueue *pq = pq_create_double();
    if (pq == NULL) return 0;
    
    pq_insert_double(pq, 3.14, 3.0);
    pq_insert_double(pq, 1.41, 1.0);
    pq_insert_double(pq, 2.71, 2.0);
    
    double val;
    double pri;
    
    pq_extract_double(pq, &val, &pri);
    if (val < 1.40 || val > 1.42 || pri != 1.0) {
        pq_free(&pq);
        return 0;
    }
    
    pq_extract_double(pq, &val, &pri);
    if (val < 2.70 || val > 2.72 || pri != 2.0) {
        pq_free(&pq);
        return 0;
    }
    
    pq_free(&pq);
    return 1;
}

int test_char_type(void) {
    PriorityQueue *pq = pq_create_char();
    if (pq == NULL) return 0;
    
    /* Insert characters with priorities */
    pq_insert_char(pq, 'c', 3.0);
    pq_insert_char(pq, 'a', 1.0);
    pq_insert_char(pq, 'b', 2.0);
    
    char val;
    pq_extract_char(pq, &val, NULL);
    if (val != 'a') {
        pq_free(&pq);
        return 0;
    }
    
    pq_extract_char(pq, &val, NULL);
    if (val != 'b') {
        pq_free(&pq);
        return 0;
    }
    
    pq_extract_char(pq, &val, NULL);
    if (val != 'c') {
        pq_free(&pq);
        return 0;
    }
    
    pq_free(&pq);
    return 1;
}

int test_large_queue(void) {
    PriorityQueue *pq = pq_create_int();
    if (pq == NULL) return 0;
    
    /* Insert 10000 elements in reverse priority order */
    for (int i = 9999; i >= 0; i--) {
        pq_insert_int(pq, i, (double)i);
    }
    
    if (pq_size(pq) != 10000) {
        pq_free(&pq);
        return 0;
    }
    
    /* Extract should be in ascending priority order */
    for (int i = 0; i < 10000; i++) {
        int val;
        double pri;
        if (!pq_extract_int(pq, &val, &pri) || val != i || pri != (double)i) {
            pq_free(&pq);
            return 0;
        }
    }
    
    int result = pq_is_empty(pq);
    pq_free(&pq);
    return result;
}

int test_null_handling(void) {
    /* These should not crash */
    pq_free(NULL);
    
    PriorityQueue *pq = NULL;
    pq_free(&pq);
    
    int val;
    double pri;
    if (pq_extract_min(NULL, &val, &pri)) return 0;
    if (pq_insert(NULL, &val, 1.0)) return 0;
    if (pq_peek_min(NULL, &val, &pri)) return 0;
    if (pq_size(NULL) != 0) return 0;
    if (!pq_is_empty(NULL)) return 0;
    
    pq = pq_create(sizeof(int), 0);
    if (pq == NULL) return 0;
    
    /* Insert with NULL data should fail */
    if (pq_insert(pq, NULL, 1.0)) {
        pq_free(&pq);
        return 0;
    }
    
    pq_free(&pq);
    return 1;
}

int test_extract_without_output(void) {
    PriorityQueue *pq = pq_create_int();
    if (pq == NULL) return 0;
    
    pq_insert_int(pq, 1, 1.0);
    pq_insert_int(pq, 2, 2.0);
    pq_insert_int(pq, 3, 3.0);
    
    /* Extract without storing */
    if (!pq_extract_min(pq, NULL, NULL) || pq_size(pq) != 2) {
        pq_free(&pq);
        return 0;
    }
    
    /* Next should be priority 2.0 */
    int val;
    double pri;
    pq_extract_int(pq, &val, &pri);
    if (val != 2 || pri != 2.0) {
        pq_free(&pq);
        return 0;
    }
    
    pq_free(&pq);
    return 1;
}

int test_struct_type(void) {
    typedef struct {
        int x;
        int y;
    } Point;
    
    PriorityQueue *pq = pq_create(sizeof(Point), 0);
    if (pq == NULL) return 0;
    
    Point p1 = {10, 20};
    Point p2 = {30, 40};
    Point p3 = {50, 60};
    
    pq_insert(pq, &p1, 3.0);
    pq_insert(pq, &p2, 1.0);  /* Highest priority */
    pq_insert(pq, &p3, 2.0);
    
    Point out;
    pq_extract_min(pq, &out, NULL);
    if (out.x != 30 || out.y != 40) {
        pq_free(&pq);
        return 0;
    }
    
    pq_extract_min(pq, &out, NULL);
    if (out.x != 50 || out.y != 60) {
        pq_free(&pq);
        return 0;
    }
    
    pq_extract_min(pq, &out, NULL);
    if (out.x != 10 || out.y != 20) {
        pq_free(&pq);
        return 0;
    }
    
    pq_free(&pq);
    return 1;
}

int test_negative_priority(void) {
    PriorityQueue *pq = pq_create_int();
    if (pq == NULL) return 0;
    
    pq_insert_int(pq, 10, 1.0);
    pq_insert_int(pq, -10, -1.0);
    pq_insert_int(pq, 0, 0.0);
    
    int val;
    double pri;
    
    pq_extract_int(pq, &val, &pri);
    if (val != -10 || pri != -1.0) {
        pq_free(&pq);
        return 0;
    }
    
    pq_extract_int(pq, &val, &pri);
    if (val != 0 || pri != 0.0) {
        pq_free(&pq);
        return 0;
    }
    
    pq_extract_int(pq, &val, &pri);
    if (val != 10 || pri != 1.0) {
        pq_free(&pq);
        return 0;
    }
    
    pq_free(&pq);
    return 1;
}

int test_priority_stability(void) {
    /* Test that heap property is maintained after many operations */
    PriorityQueue *pq = pq_create_int();
    if (pq == NULL) return 0;
    
    /* Insert many elements */
    for (int i = 0; i < 100; i++) {
        pq_insert_int(pq, i, (double)(rand() % 1000));
    }
    
    /* Extract all and verify min-heap property */
    double last_priority = -1.0;
    while (!pq_is_empty(pq)) {
        int val;
        double pri;
        pq_extract_int(pq, &val, &pri);
        if (pri < last_priority) {
            pq_free(&pq);
            return 0;  /* Priority order violated */
        }
        last_priority = pri;
    }
    
    pq_free(&pq);
    return 1;
}

/* ========== Main ========== */

int main(void) {
    printf("\n========================================\n");
    printf("  Priority Queue Utils Test Suite\n");
    printf("========================================\n\n");
    
    /* Basic operations */
    TEST(create_free);
    TEST(create_with_capacity);
    TEST(insert_extract_int);
    TEST(peek);
    TEST(same_priority);
    TEST(is_empty);
    TEST(clear);
    
    /* Memory management */
    TEST(reserve);
    TEST(shrink_to_fit);
    
    /* Advanced operations */
    TEST(contains_priority);
    TEST(update_priority);
    TEST(merge);
    TEST(get_sorted);
    
    /* Different types */
    TEST(double_type);
    TEST(char_type);
    TEST(struct_type);
    
    /* Edge cases */
    TEST(negative_priority);
    TEST(priority_stability);
    TEST(null_handling);
    TEST(extract_without_output);
    
    /* Stress test */
    TEST(large_queue);
    
    printf("\n========================================\n");
    printf("  Results: %d/%d tests passed\n", tests_passed, tests_run);
    printf("========================================\n\n");
    
    return (tests_passed == tests_run) ? 0 : 1;
}

File: Lua/priority_queue_utils/test_priority_queue.lua
--- Test suite for PriorityQueue
--- Run with: lua test_priority_queue.lua

local PriorityQueue = dofile("priority_queue.lua")

local tests_passed = 0
local tests_failed = 0

local function test(name, fn)
    local ok, err = pcall(fn)
    if ok then
        tests_passed = tests_passed + 1
        print("✓ " .. name)
    else
        tests_failed = tests_failed + 1
        print("✗ " .. name .. ": " .. tostring(err))
    end
end

local function assert_equal(a, b, msg)
    if a ~= b then
        error(msg .. " (expected: " .. tostring(b) .. ", got: " .. tostring(a) .. ")")
    end
end

local function assert_true(value, msg)
    if not value then
        error(msg .. " (expected true, got false)")
    end
end

local function assert_false(value, msg)
    if value then
        error(msg .. " (expected false, got true)")
    end
end

print("=== PriorityQueue Tests ===\n")

-- Test: Create max-heap
test("Create max-heap priority queue", function()
    local pq = PriorityQueue.max_heap()
    assert_true(pq:is_empty(), "Should be empty initially")
    assert_equal(pq:len(), 0, "Length should be 0")
end)

-- Test: Create min-heap
test("Create min-heap priority queue", function()
    local pq = PriorityQueue.min_heap()
    assert_true(pq:is_empty(), "Should be empty initially")
end)

-- Test: Push and pop single element
test("Push and pop single element", function()
    local pq = PriorityQueue.max_heap()
    pq:push("task1", 5)
    assert_equal(pq:len(), 1, "Length should be 1")
    assert_false(pq:is_empty(), "Should not be empty")
    
    local value, priority = pq:pop()
    assert_equal(value, "task1", "Value should be task1")
    assert_equal(priority, 5, "Priority should be 5")
    assert_true(pq:is_empty(), "Should be empty after pop")
end)

-- Test: Max-heap ordering
test("Max-heap ordering (highest priority first)", function()
    local pq = PriorityQueue.max_heap()
    pq:push("low", 1)
    pq:push("high", 10)
    pq:push("medium", 5)
    
    local v1, p1 = pq:pop()
    assert_equal(v1, "high", "First should be high priority")
    assert_equal(p1, 10, "Priority should be 10")
    
    local v2, p2 = pq:pop()
    assert_equal(v2, "medium", "Second should be medium priority")
    assert_equal(p2, 5, "Priority should be 5")
    
    local v3, p3 = pq:pop()
    assert_equal(v3, "low", "Third should be low priority")
    assert_equal(p3, 1, "Priority should be 1")
end)

-- Test: Min-heap ordering
test("Min-heap ordering (lowest priority first)", function()
    local pq = PriorityQueue.min_heap()
    pq:push("low", 1)
    pq:push("high", 10)
    pq:push("medium", 5)
    
    local v1, p1 = pq:pop()
    assert_equal(v1, "low", "First should be low priority")
    assert_equal(p1, 1, "Priority should be 1")
    
    local v2, p2 = pq:pop()
    assert_equal(v2, "medium", "Second should be medium priority")
    assert_equal(p2, 5, "Priority should be 5")
    
    local v3, p3 = pq:pop()
    assert_equal(v3, "high", "Third should be high priority")
    assert_equal(p3, 10, "Priority should be 10")
end)

-- Test: Peek
test("Peek returns top without removing", function()
    local pq = PriorityQueue.max_heap()
    pq:push("task1", 5)
    pq:push("task2", 10)
    
    local value, priority = pq:peek()
    assert_equal(value, "task2", "Should peek highest priority")
    assert_equal(priority, 10, "Priority should be 10")
    assert_equal(pq:len(), 2, "Length should still be 2")
    
    pq:pop()
    value, priority = pq:peek()
    assert_equal(value, "task1", "Should peek next element")
    assert_equal(priority, 5, "Priority should be 5")
end)

-- Test: Peek on empty queue
test("Peek on empty queue returns nil", function()
    local pq = PriorityQueue.max_heap()
    local value, priority = pq:peek()
    assert_equal(value, nil, "Value should be nil")
    assert_equal(priority, nil, "Priority should be nil")
end)

-- Test: Pop on empty queue
test("Pop on empty queue returns nil", function()
    local pq = PriorityQueue.max_heap()
    local value, priority = pq:pop()
    assert_equal(value, nil, "Value should be nil")
    assert_equal(priority, nil, "Priority should be nil")
end)

-- Test: Enqueue/dequeue aliases
test("Enqueue/dequeue aliases work correctly", function()
    local pq = PriorityQueue.max_heap()
    pq:enqueue("item", 5)
    assert_equal(pq:len(), 1, "Length should be 1")
    
    local value, priority = pq:dequeue()
    assert_equal(value, "item", "Value should be item")
    assert_equal(priority, 5, "Priority should be 5")
end)

-- Test: Clear
test("Clear removes all elements", function()
    local pq = PriorityQueue.max_heap()
    pq:push("a", 1)
    pq:push("b", 2)
    pq:push("c", 3)
    assert_equal(pq:len(), 3, "Length should be 3")
    
    pq:clear()
    assert_equal(pq:len(), 0, "Length should be 0 after clear")
    assert_true(pq:is_empty(), "Should be empty after clear")
end)

-- Test: Contains
test("Contains finds existing elements", function()
    local pq = PriorityQueue.max_heap()
    pq:push("apple", 1)
    pq:push("banana", 2)
    pq:push("cherry", 3)
    
    assert_true(pq:contains("apple"), "Should find apple")
    assert_true(pq:contains("banana"), "Should find banana")
    assert_false(pq:contains("grape"), "Should not find grape")
end)

-- Test: Contains with custom equality
test("Contains with custom equality function", function()
    local pq = PriorityQueue.max_heap()
    pq:push({ id = 1, name = "Alice" }, 5)
    pq:push({ id = 2, name = "Bob" }, 3)
    
    local custom_equals = function(a, b) return a.id == b.id end
    assert_true(pq:contains({ id = 1 }, custom_equals), "Should find by id")
    assert_false(pq:contains({ id = 3 }, custom_equals), "Should not find missing id")
end)

-- Test: Remove
test("Remove specific element", function()
    local pq = PriorityQueue.max_heap()
    pq:push("a", 1)
    pq:push("b", 2)
    pq:push("c", 3)
    
    assert_true(pq:remove("b"), "Should remove b")
    assert_equal(pq:len(), 2, "Length should be 2")
    assert_false(pq:contains("b"), "b should be removed")
    
    local v1 = pq:pop()
    local v2 = pq:pop()
    assert_true((v1 == "a" or v1 == "c") and (v2 == "a" or v2 == "c"), "Remaining elements should be a and c")
end)

-- Test: Remove non-existent
test("Remove non-existent element returns false", function()
    local pq = PriorityQueue.max_heap()
    pq:push("a", 1)
    assert_false(pq:remove("x"), "Should return false for non-existent")
end)

-- Test: Update priority
test("Update priority of existing element", function()
    local pq = PriorityQueue.max_heap()
    pq:push("task", 5)
    pq:push("other", 3)
    
    pq:update_priority("task", 10)
    local value, priority = pq:peek()
    assert_equal(value, "task", "task should now be at top")
    assert_equal(priority, 10, "Priority should be 10")
end)

-- Test: Update priority (decrease)
test("Update priority to lower value", function()
    local pq = PriorityQueue.max_heap()
    pq:push("high", 10)
    pq:push("low", 1)
    
    pq:update_priority("high", 0)
    local value, priority = pq:peek()
    assert_equal(value, "low", "low should now be at top")
    assert_equal(priority, 1, "Priority should be 1")
end)

-- Test: Update priority with custom equality
test("Update priority with custom equality", function()
    local pq = PriorityQueue.max_heap()
    pq:push({ id = 1, name = "Alice" }, 5)
    pq:push({ id = 2, name = "Bob" }, 3)
    
    local custom_equals = function(a, b) return a.id == b.id end
    assert_true(pq:update_priority({ id = 1 }, 10, custom_equals), "Should update by id")
    
    local top, _ = pq:peek()
    assert_equal(top.id, 1, "Alice should be at top")
end)

-- Test: Merge
test("Merge two priority queues", function()
    local pq1 = PriorityQueue.max_heap()
    pq1:push("a", 5)
    pq1:push("b", 3)
    
    local pq2 = PriorityQueue.max_heap()
    pq2:push("c", 10)
    pq2:push("d", 1)
    
    pq1:merge(pq2)
    assert_equal(pq1:len(), 4, "Length should be 4")
    
    local values = {}
    while not pq1:is_empty() do
        local v, _ = pq1:pop()
        table.insert(values, v)
    end
    assert_equal(values[1], "c", "First should be c (priority 10)")
    assert_equal(values[2], "a", "Second should be a (priority 5)")
    assert_equal(values[3], "b", "Third should be b (priority 3)")
    assert_equal(values[4], "d", "Fourth should be d (priority 1)")
end)

-- Test: To sorted array
test("To sorted array returns elements in order", function()
    local pq = PriorityQueue.max_heap()
    pq:push("low", 1)
    pq:push("high", 10)
    pq:push("medium", 5)
    
    local sorted = pq:to_sorted_array()
    assert_equal(#sorted, 3, "Should have 3 elements")
    assert_equal(sorted[1].value, "high", "First should be high")
    assert_equal(sorted[2].value, "medium", "Second should be medium")
    assert_equal(sorted[3].value, "low", "Third should be low")
    
    -- Original queue should not be modified
    assert_equal(pq:len(), 3, "Original queue should still have 3 elements")
end)

-- Test: Custom comparator
test("Custom comparator (reverse alphabetical)", function()
    local pq = PriorityQueue.with_comparator(function(a, b)
        return a > b -- Higher string value first (reverse alphabetical)
    end)
    
    -- Use string itself as priority (for custom ordering demonstration)
    pq:push("apple", "apple")
    pq:push("banana", "banana")
    pq:push("cherry", "cherry")
    
    assert_equal(pq:pop(), "cherry", "cherry should come first")
    assert_equal(pq:pop(), "banana", "banana should come second")
    assert_equal(pq:pop(), "apple", "apple should come last")
end)

-- Test: Large number of elements
test("Handle many elements correctly", function()
    local pq = PriorityQueue.min_heap()
    local count = 100
    
    for i = count, 1, -1 do
        pq:push("item" .. i, i)
    end
    
    assert_equal(pq:len(), count, "Should have " .. count .. " elements")
    
    for i = 1, count do
        local value, priority = pq:pop()
        assert_equal(priority, i, "Should pop in ascending order")
    end
    
    assert_true(pq:is_empty(), "Should be empty after all pops")
end)

-- Test: Same priority handling (stable-ish behavior)
test("Elements with same priority", function()
    local pq = PriorityQueue.max_heap()
    pq:push("first", 5)
    pq:push("second", 5)
    pq:push("third", 5)
    
    assert_equal(pq:len(), 3, "Should have 3 elements")
    
    -- All should have priority 5
    local _, p1 = pq:pop()
    local _, p2 = pq:pop()
    local _, p3 = pq:pop()
    
    assert_equal(p1, 5, "Priority should be 5")
    assert_equal(p2, 5, "Priority should be 5")
    assert_equal(p3, 5, "Priority should be 5")
end)

-- Test: Method chaining
test("Method chaining on push", function()
    local pq = PriorityQueue.max_heap()
        :push("a", 1)
        :push("b", 2)
        :push("c", 3)
    
    assert_equal(pq:len(), 3, "Should have 3 elements")
end)

print("\n=== Test Results ===")
print(string.format("Passed: %d", tests_passed))
print(string.format("Failed: %d", tests_failed))

if tests_failed == 0 then
    print("\n✓ All tests passed!")
else
    print("\n✗ Some tests failed.")
    os.exit(1)
end